add typespecs for errors

ericsullivan opened the pull request:
This addresses https://github.com/ueberauth/guardian/issues/253. It includes https://github.com/ueberauth/guardian/pull/255, but I wasn't sure if I got the error types correct. 

I think the error type can be an atom or string. In GuardianDB, it looks like it's always an atom, but the typespec for methods that call the hooks allow both.

```
  @spec decode_and_verify(String.t, map) :: {:ok, map} |
                                            {:error, atom | String.t}

  @spec encode_and_sign(any, atom | String.t, map) :: {:ok, String.t, map} |
                                                      {:error, atom} |
                                                      {:error, String.t}
```

Guardian.Hooks.Test also returns strings, but I wrote that, so can fix it if it should be atoms... Did you want to keep it being both or only {:error, atom}?

zachdaniel commented:
:+1:

asummers commented:
Thank you for addressing my issue @ericsullivan!

asummers commented:
Thinking about this for a bit, even if Guardian only supports `atom`, it should probably allow `any`. If you're trying to use a hook to return some custom error, say an `Error` struct, having this be `any` would allow the most future flexibility.

doomspork commented:
That's a good point @asummers, @ericsullivan would you mind incorporating that change?

ericsullivan commented:
I added {:error, any}. Then for the methods that returned the results of calling the hooks I updated them. Then some methods that called them... I'm not confident I got them all, but dialyzer passed. This seems like a larger change than I intended. Would it be best to fix the original issue in this PR, then open another one to support any?

asummers commented:
I don't see any value in splitting that work, personally. Those changes are all of the "boring" variety, so it's unlikely to break something. I think it looks great as is :)

doomspork commented:
Thank you @ericsullivan! 👍
